doom/.doom.d/config.org
#+TITLE: Doom Emacs configuration
#+AUTHOR: Marco Pas
#+EMAIL: marco.pasopas@gmail.com
#+EXPORT_FILE_NAME: README.md
#+PROPERTY: header-args:emacs-lisp :tangle config.el
#+STARTUP: fold

* Table of Contents                                                                                             :TOC_3:noexport:
- [[#config-file-header][Config file header]]
- [[#basic-settings][Basic settings]]
  - [[#directories][Directories]]
  - [[#personal-information][Personal Information]]
- [[#general-configuration][General configuration]]
  - [[#set-the-authentication-source-to-use][Set the authentication source to use]]
  - [[#disable-exit-confirmation][Disable exit confirmation]]
  - [[#enable-line-numbers][Enable line numbers]]
  - [[#prompt-for-buffer-after-a-split][Prompt for buffer after a split]]
  - [[#set-to-open-the-correct-application-to-open-a-file-link][Set to open the correct application to open a file link]]
- [[#custom-functions][Custom functions]]
  - [[#insert-date][Insert date]]
  - [[#insert-a-timestamp][Insert a timestamp]]
  - [[#find-files-that-contain-a-specific-text-recursive][Find files that contain a specific text (recursive)]]
  - [[#convert-a-region-from-markdown-to-org][Convert a region from markdown to org]]
- [[#keybindings][Keybindings]]
  - [[#remove-entire-line-when-at-the-beginning-c-k][Remove entire line when at the beginning =C-k=]]
  - [[#cycle-through-the-open-windows-c-w-c-w][Cycle through the open windows =C-w C-w=]]
  - [[#grab-open-link-from-application-c-c-g][Grab open link from application =C-c g=]]
  - [[#toggle-truncate-lines-spc-t-t][Toggle Truncate lines =SPC t t=]]
  - [[#open-files-from-the-bookmarks-list-c-c-b][Open files from the bookmarks list =C-c b=]]
  - [[#openfocus-on-iterm-spc-f-i][Open/focus on iTerm =SPC f i=]]
  - [[#toggle-pretty-markup-c-c-e][Toggle pretty markup =C-c e=]]
  - [[#remove-all-results-from-an-org-document-s-][Remove all results from an org document =s-]=]]
  - [[#delete-current-buffer-and-file-c-c-d][Delete current buffer and file =C-c D=]]
  - [[#insert-clipboard-image-c-m-y][Insert clipboard image =C-M-y=]]
- [[#visual-settings][Visual settings]]
  - [[#maximize-emacs-frame-on-startup][Maximize Emacs frame on startup]]
  - [[#theme][Theme]]
  - [[#fonts-and-icons][Fonts and Icons]]
    - [[#default-font][Default font]]
    - [[#variable-pitch][Variable pitch]]
    - [[#big-font][Big font]]
    - [[#enable-bolditalic-fonts][Enable bold/italic fonts]]
    - [[#set-custom-faces][Set custom faces]]
  - [[#modeline][Modeline]]
    - [[#limit-mode-line-vcs-length][Limit mode line vcs length]]
    - [[#enable-the-display-of-time-and-date-in-the-mode-line][Enable the display of time and date in the mode line]]
- [[#package-configuration][Package configuration]]
  - [[#auto-dim-other-buffers][Auto dim other buffers]]
  - [[#centaur-tab][Centaur tab]]
  - [[#company-mode][Company mode]]
  - [[#deft][Deft]]
  - [[#dired][Dired]]
    - [[#enable-dual-pane][Enable dual pane]]
  - [[#forge][Forge]]
  - [[#magit][Magit]]
  - [[#ob-mermaid][Ob-mermaid]]
  - [[#org-mode][Org mode]]
  - [[#org-agenda][Org agenda]]
  - [[#org-calendar][Org calendar]]
  - [[#org-capture-templates][Org capture templates]]
  - [[#org-babel][Org babel]]
  - [[#org-roam][Org roam]]
  - [[#org-tree-slide][Org tree slide]]
  - [[#zoom-window][Zoom window]]
  - [[#smartparens][Smartparens]]
- [[#programming][Programming]]
  - [[#css][CSS]]
  - [[#json][JSON]]
    - [[#c-c-c-f-json-beautify][=C-c C-f= Json beautify]]
  - [[#markdown][Markdown]]
  - [[#clojure--cider][Clojure / Cider]]

* Config file header

#+html: <details><summary>config.el</summary>
#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; DO NOT EDIT THIS FILE DIRECTLY
;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!
#+end_src
#+html: </details>

* Basic settings
** Directories

Base directories that are used for org/roam/deft

#+begin_src emacs-lisp
(setq my/home-dir "~")
(setq my/org-base-dir (concat my/home-dir "/Dropbox/notes/org/"))
(setq my/org-roam-dir (concat my/org-base-dir "roam"))

;; set the default directories for org/roam
(setq org-directory my/org-base-dir)
(setq org-roam-directory my/org-roam-dir)
(setq deft-directory my/org-base-dir)
#+end_src

** Personal Information

Let's set some variables with basic user information.

#+BEGIN_SRC emacs-lisp
(setq user-full-name "Marco Pas")
(setq user-mail-address "marco.pasopas@gmail.com")
#+END_SRC

* General configuration
** Set the authentication source to use

#+begin_src emacs-lisp
(setq auth-sources '("~/.authinfo"))
#+end_src

** Disable exit confirmation

#+begin_src emacs-lisp
(setq confirm-kill-emacs nil)
#+end_src

** Enable line numbers

#+begin_src emacs-lisp
(setq display-line-numbers-type t)
#+end_src

** Prompt for buffer after a split

#+begin_src emacs-lisp
(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (consult-buffer))
#+end_src

** Set to open the correct application to open a file link

#+begin_src emacs-lisp
(setq org-file-apps
      '(("\\.docx\\'" . default)
        ("\\.x?html?\\'" . default)
        (auto-mode . emacs)))
#+end_src

* Custom functions
** Insert date

#+begin_src emacs-lisp
(defun my/insert-date (arg)
  (interactive "P")
  (insert (if arg
              (format-time-string "%d.%m.%Y")
            (format-time-string "%Y-%m-%d"))))
#+end_src

** Insert a timestamp

#+begin_src emacs-lisp
(defun my/insert-timestamp ()
  (interactive)
  (insert (format-time-string "%Y-%m-%dT%H:%M:%S")))
#+end_src

** Find files that contain a specific text (recursive)

I use this function in org-agenda that fills the variable with files that have a required text in it.

#+begin_src emacs-lisp
(defun my/find-files-that-contain (include-dir text &optional with-extension exclude-dir)
  (let ((shell-command
         (concat
          "grep --recursive --files-with-matches "
          (when with-extension (concat "--include=\"" with-extension "\" "))
          (when exclude-dir (concat "--exclude-dir=\"" exclude-dir "\" "))
          "\"" (shell-quote-argument text) "\" "
          include-dir)))
    (split-string (shell-command-to-string shell-command))))
#+end_src

** Convert a region from markdown to org

#+begin_src emacs-lisp
(defun my/md-to-org-region (start end)
  "Convert region from markdown to org"
  (interactive "r")
  (shell-command-on-region start end "pandoc -f markdown -t org" t t))
#+end_src

* Keybindings
** Remove entire line when at the beginning =C-k=

When at the beginning of the line, make =Ctrl-K= remove the whole line, instead of just emptying it.

#+begin_src emacs-lisp
(setq kill-whole-line t)
#+end_src

** Cycle through the open windows =C-w C-w=

#+begin_src emacs-lisp
(map! :map evil-normal-state-map
      :desc "Cycle through windows"
      :ng "C-w C-w" #'evil-window-next)
#+end_src

** Grab open link from application =C-c g=

Shortcut to grab a link from an open Mac application.

#+begin_src emacs-lisp
(add-hook 'org-mode-hook (lambda ()
                           (define-key org-mode-map (kbd "C-c g") 'org-mac-grab-link)))
#+end_src

** Toggle Truncate lines =SPC t t=

#+begin_src emacs-lisp
(map! :leader
      :desc "Toggle truncate lines"
      "t t" #'toggle-truncate-lines)
#+end_src

** Open files from the bookmarks list =C-c b=

#+begin_src emacs-lisp
(map! :desc "Open bookmarks list"
      :n "C-c b" 'list-bookmarks)
#+end_src

** Open/focus on iTerm =SPC f i=

#+begin_src emacs-lisp
(defun my/iterm-focus ()
  (interactive)
  (do-applescript
   " do shell script \"open -a iTerm\"\n"))

(map! :leader
      :desc "Open/focus iTerm"
      :"f i" #'my/iterm-focus)
#+end_src

** Toggle pretty markup =C-c e=

#+begin_src emacs-lisp
(after! org
  ;; (define-key org-mode-map (kbd "C-c e") '+org-pretty-mode))
  (define-key org-mode-map (kbd "C-c e") 'font-lock-mode))
#+end_src

** Remove all results from an org document =s-]=

#+begin_src emacs-lisp
(defconst my/org-special-pre "^\s*#[+]")

(defun my/org-2every-src-block (fn)
  "Visit every Source-Block and evaluate `FN'."
  (interactive)
  (save-excursion
    (goto-char (point-min))
    (let ((case-fold-search t))
      (while (re-search-forward (concat my/org-special-pre "BEGIN_SRC") nil t)
        (let ((element (org-element-at-point)))
          (when (eq (org-element-type element) 'src-block)
            (funcall fn element)))))
    (save-buffer)))

(after! org
   (define-key org-mode-map (kbd "s-]") (lambda () (interactive)
                                       (my/org-2every-src-block
                                        'org-babel-remove-result))))
#+end_src

** Delete current buffer and file =C-c D=

#+begin_src emacs-lisp
(defun er-delete-file-and-buffer ()
  "Kill the current buffer and deletes the file it is visiting."
  (interactive)
  (let ((filename (buffer-file-name)))
    (when filename
      (if (vc-backend filename)
          (vc-delete-file filename)
        (progn
          (delete-file filename)
          (message "Deleted file %s" filename)
          (kill-buffer))))))


(global-set-key (kbd "C-c D")  #'er-delete-file-and-buffer)
#+end_src

** Insert clipboard image =C-M-y=

Use the current headername as part of the image filename
#+begin_src emacs-lisp
(setq my/insert-clipboard-image-use-headername t)
#+end_src

Use the current buffername as part of the screenshot filename
#+begin_src emacs-lisp
(setq my/insert-clipboard-image-use-buffername t)
#+end_src

Function that allows a screenshot to be pasted in a Org-mode file. The relative link will be pasted into the document and a relative =./images= will be created to store the images.

#+begin_src emacs-lisp
;; Overview
;; --------
;; Inserts an image from the clipboard by prompting the user for a filename.
;; Default extension for the pasted filename is .png

;; A ./images directory will be created relative to the current Org-mode document to store the images.

;; The default name format of the pasted image is:
;; filename: <yyyymmdd>_<hhmmss>_-_<image-filename>.png

;; Important
;; --------
;; This function depends on 'pngpaste' to paste the clipboard image
;; -> $ brew install pngpaste

;; Basic Customization
;; -------------------
;; Include the current Org-mode header as part of the image name.
;; (setq my/insert-clipboard-image-use-headername t)
;; filename: <yyyymmdd>_<hhmmss>_-_<headername>_-_<image-filename>.png

;; Include the buffername as part of the image name.
;; (setq my/insert-clipboard-image-use-buffername t)
;; filename: <yyyymmdd>_<hhmmss>_-_<buffername>_-_<image-filename>.png

;; Full name format
;; filename: <yyyymmdd>_<hhmmss>_-_<buffername>_-_<headername>_-_<image-filename>.png
(defun my/insert-clipboard-image (filename)
  "Inserts an image from the clipboard"
  (interactive "sFilename to paste: ")
  (let ((file
         (concat
          (file-name-directory buffer-file-name)
          "images/"
          (format-time-string "%Y%m%d_%H%M%S_-_")
          (if (bound-and-true-p my/insert-clipboard-image-use-buffername)
              (concat (s-replace "-" "_"
                                 (downcase (file-name-sans-extension (buffer-name)))) "_-_")
            "")
          (if (bound-and-true-p my/insert-clipboard-image-use-headername)
              (concat (s-replace " " "_" (downcase (nth 4 (org-heading-components)))) "_-_")
            "")
          filename ".png")))

    ;; create images directory
    (unless (file-exists-p (file-name-directory file))
      (make-directory (file-name-directory file)))

    ;; paste file from clipboard
    (print file)
    (shell-command (concat "pngpaste " "\"" file "\""))
    (insert (concat "[[./images/" (file-name-nondirectory file) "]]"))))
#+end_src

Bind insert clipboard image to Ctrl-Option-y

#+begin_src emacs-lisp
(map! :desc "Insert clipboard image"
      :n "C-M-y" 'my/insert-clipboard-image)
#+end_src

* Visual settings
** Maximize Emacs frame on startup

#+begin_src emacs-lisp
(add-to-list 'initial-frame-alist '(fullscreen . maximized))
#+end_src

** Theme

Sets the default theme ; currently doom-snazzy

#+BEGIN_SRC emacs-lisp
(setq doom-theme 'doom-snazzy)
#+end_src

** Fonts and Icons
*** Default font

doom-font is the default font for emacs to use

#+begin_src emacs-lisp
(setq doom-font (font-spec :family "JetBrains Mono" :weight 'light :size 16))
#+end_src

*** Variable pitch

doom-variable-pitch-font is a proportional font used for reading emails, doc etc.

#+begin_src emacs-lisp
(setq doom-variable-pitch-font (font-spec :family "JetBrains Mono" :weight 'light :size 16))
#+end_src

*** Big font

doom-big-font is the default font to use for doom-big-font-mode

#+begin_src emacs-lisp
(setq doom-big-font (font-spec :family "JetBrains Mono" :size 30))
#+end_src

*** Enable bold/italic fonts

#+begin_src emacs-lisp
(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))
#+end_src

*** Set custom faces

#+begin_src emacs-lisp
(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))
#+end_src

** Modeline
*** Limit mode line vcs length

#+begin_src emacs-lisp
(setq doom-modeline-vcs-max-length 80)
#+end_src

#+begin_src emacs-lisp
(after! doom-modeline
  (setq doom-modeline-bar-width 3))
#+end_src

*** Enable the display of time and date in the mode line

#+begin_src emacs-lisp
(display-time-mode 1)
(setq display-time-day-and-date t)
#+end_src

* Package configuration
** Auto dim other buffers

The auto-dim-other-buffers-mode is a global minor mode which makes windows without focus less prominent. With many windows in a frame, the idea is that this mode helps recognize which is the selected window by providing a non-intrusive but still noticeable visual indicator.

#+begin_src emacs-lisp
(add-hook 'after-init-hook (lambda ()
  (when (fboundp 'auto-dim-other-buffers-mode)
    (auto-dim-other-buffers-mode t))))
#+end_src

** Centaur tab

#+begin_src emacs-lisp
(define-key evil-normal-state-map (kbd "g t") 'centaur-tabs-forward)
(define-key evil-normal-state-map (kbd "g T") 'centaur-tabs-backward)
(define-key evil-normal-state-map (kbd "C-<prior>") 'centaur-tabs-backward)
(define-key evil-normal-state-map (kbd "C-<next>") 'centaur-tabs-forward)
(define-key evil-normal-state-map (kbd "C-c t s") 'centaur-tabs-counsel-switch-group)
(define-key evil-normal-state-map (kbd "C-c t p") 'centaur-tabs-group-by-projectile-project)
(define-key evil-normal-state-map (kbd "C-c t g") 'centaur-tabs-group-buffer-groups)
#+end_src

** Company mode

#+begin_src emacs-lisp
;; (set-company-backend! 'org-mode '(:separate company-yasnippet company-dabbrev))

;; (setq +company-backend-alist (assq-delete-all 'text-mode +company-backend-alist))
;; (add-to-list '+company-backend-alist '(text-mode (:separate company-dabbrev company-yasnippet)))

(after! company
  (set-company-backend! 'org-mode '(:separate company-yasnippet company-dabbrev)))

(after! company
  (setq company-idle-delay 0.5 ;; How long to wait before popping up
        company-minimum-prefix-length 1 ;; Show the menu after one key press
        company-tooltip-limit 15 ;; Limit on how many options to display
        company-selection-wrap-around t ;; after end of list warap around to start
        company-tooltip-align-annotations t ;; Align annotations to the right
        company-require-match nil           ;; Allow free typing
        company-selection-wrap-around t ;; Wrap around to beginning when you hit bottom of suggestions
        company-dabbrev-ignore-case t ;; Don't ignore case when completing
        company-dabbrev-downcase nil ;; Don't automatically downcase completions
        ))
#+end_src

** Deft

#+begin_src emacs-lisp
(setq deft-default-extension "org"
      deft-extensions '("txt" "md" "org")
      deft-recursive t
      deft-use-filename-as-title t
      deft-use-filter-string-for-filename t
      deft-file-naming-rules '((noslash . "-")
                               (nospace . "-")
                               (case-fn . downcase)))
#+end_src

** Dired

*** Enable dual pane


doom/.doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; DO NOT EDIT THIS FILE DIRECTLY
;; This is a file generated from a literate programing source file located at
;; https://gitlab.com/zzamboni/dot-doom/-/blob/master/doom.org
;; You should make any changes there and regenerate it from Emacs org-mode
;; using org-babel-tangle (C-c C-v t)

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!

;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
;; (setq user-full-name "John Doe"
;;      user-mail-address "john@doe.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 13))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
;; (setq doom-theme 'doom-one)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
;; (setq org-directory "~/org/")

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
;; (setq display-line-numbers-type t)

;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(setq my/home-dir "~")
(setq my/org-base-dir (concat my/home-dir "/Dropbox/notes/org/"))

(setq user-full-name "Marco Pas")

(setq user-mail-address "marco.pasopas@gmail.com")

(setq auth-sources '("~/.authinfo"))

(setq confirm-kill-emacs nil)

(setq kill-whole-line t)

(map! :map evil-normal-state-map
      :desc "Close the current buffer"
      :ng "c" #'kill-this-buffer)

(map! :map evil-normal-state-map
      :desc "Cycle through windows"
      :ng "C-w C-w" #'evil-window-next)

(add-hook 'org-mode-hook (lambda ()
                           (define-key org-mode-map (kbd "C-c g") 'org-mac-grab-link)))

(defun insert-date (arg)
  (interactive "P")
  (insert (if arg
              (format-time-string "%d.%m.%Y")
            (format-time-string "%Y-%m-%d"))))

(defun insert-timestamp ()
  (interactive)
  (insert (format-time-string "%Y-%m-%dT%H:%M:%S")))

(defun jx/insert-custom-clock-entry ()
  (interactive)
  (insert "CLOCK: ")
  (org-time-stamp-inactive)
  (insert "--")
  ;; Inserts the current time by default.
  (let ((current-prefix-arg '(4))) (call-interactively 'org-time-stamp-inactive))
  (org-ctrl-c-ctrl-c))

                                        ; Shortcut
(define-key evil-normal-state-map (kbd "C-'") 'jx/insert-custom-clock-entry)
(define-key evil-insert-state-map (kbd "C-'") 'jx/insert-custom-clock-entry)

(setq my/insert-clipboard-image-use-headername t)

(setq my/insert-clipboard-image-use-buffername t)

;; Overview
;; --------
;; Inserts an image from the clipboard by prompting the user for a filename.
;; Default extension for the pasted filename is .png

;; A ./images directory will be created relative to the current Org-mode document to store the images.

;; The default name format of the pasted image is:
;; filename: <yyyymmdd>_<hhmmss>_-_<image-filename>.png

;; Important
;; --------
;; This function depends on 'pngpaste' to paste the clipboard image
;; -> $ brew install pngpaste

;; Basic Customization
;; -------------------
;; Include the current Org-mode header as part of the image name.
;; (setq my/insert-clipboard-image-use-headername t)
;; filename: <yyyymmdd>_<hhmmss>_-_<headername>_-_<image-filename>.png

;; Include the buffername as part of the image name.
;; (setq my/insert-clipboard-image-use-buffername t)
;; filename: <yyyymmdd>_<hhmmss>_-_<buffername>_-_<image-filename>.png

;; Full name format
;; filename: <yyyymmdd>_<hhmmss>_-_<buffername>_-_<headername>_-_<image-filename>.png
(defun my/insert-clipboard-image (filename)
  "Inserts an image from the clipboard"
  (interactive "sFilename to paste: ")
  (let ((file
         (concat
          (file-name-directory buffer-file-name)
          "images/"
          (format-time-string "%Y%m%d_%H%M%S_-_")
          (if (bound-and-true-p my/insert-clipboard-image-use-buffername)
              (concat (s-replace "-" "_"
                                 (downcase (file-name-sans-extension (buffer-name)))) "_-_")
            "")
          (if (bound-and-true-p my/insert-clipboard-image-use-headername)
              (concat (s-replace " " "_" (downcase (nth 4 (org-heading-components)))) "_-_")
            "")
          filename ".png")))

    ;; create images directory
    (unless (file-exists-p (file-name-directory file))
      (make-directory (file-name-directory file)))

    ;; paste file from clipboard
    (shell-command (concat "pngpaste " file))
    (insert (concat "[[./images/" (file-name-nondirectory file) "]]"))))

(map! :desc "Insert clipboard image"
      :n "C-M-y" 'my/insert-clipboard-image)

(add-to-list 'initial-frame-alist '(fullscreen . maximized))

(setq doom-theme 'doom-vibrant)
(setq doom-vibrant-brighter-comments t  ;; "If non-nil, more vivid colors will be used to style the mode-line."
      doom-vibrant-brighter-modeline t  ;; "If non-nil, comments will be highlighted in more vivid colors."
      doom-vibrant-comment-bg t)        ;; "If non-nil, comments will have a subtle, darker background. Enhancing their legibility."

(setq doom-font (font-spec :family "FiraCode Nerd Font Mono" :size 16))

(setq doom-variable-pitch-font (font-spec :family "Ubuntu" :size 14))

(setq doom-big-font (font-spec :family "FiraCode Nerd Font Mono" :size 30))

(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))

(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)

(map! :leader
      :desc "Toggle truncate lines"
      "t t" #'toggle-truncate-lines)

(display-time-mode 1)
(setq display-time-day-and-date t)

(setq doom-modeline-vcs-max-length 80)

(after! doom-modeline
  (setq doom-modeline-bar-width 3))

(add-hook 'after-init-hook (lambda ()
  (when (fboundp 'auto-dim-other-buffers-mode)
    (auto-dim-other-buffers-mode t))))

(setq deft-directory my/org-base-dir
      deft-default-extension "org"
      deft-extensions '("txt" "md" "org")
      deft-recursive t
      deft-use-filename-as-title t
      deft-use-filter-string-for-filename t
      deft-file-naming-rules '((noslash . "-")
                               (nospace . "-")
                               (case-fn . downcase)))

(setq dired-dwim-target t)

;; (with-eval-after-load 'forge
;; Configure number of topics show, open and closed
;; use negative number to toggle the view of closed topics
;; using `forge-toggle-closed-visibility'
(setq  forge-topic-list-limit '(100 . -10))

;; set closed to 0 to never show closed issues
;; (setq  forge-topic-list-limit '(100 . 0))

(defun acg/parse-git-remote-url (url)
  "If necessary, convert an SSH to HTTPS git remote location."
  (if (string-match "^http" url)
      url
    (replace-regexp-in-string "\\(.*\\)@\\(.*\\):\\(.*\\)\\(\\.git?\\)"
                              "https://\\2/\\3"
                              url)))

(defun acg/magit-open-remote-repo ()
  "Opens a remote repo URL. Prompts the user to choose a remote."
  (interactive)
  (let* ((remote-name (magit-read-remote "Choose remote repository"))
         (url (magit-get "remote" remote-name "url")))
    (browse-url (acg/parse-git-remote-url url))
    (message "Opening repo %s" url)))

(defun acg/project-get-root-relative-path (&optional path)
  "Returns PATH relative to a project root."
  (let* ((path (expand-file-name (or path default-directory)))
         (root-path (expand-file-name
                     (cdr (project--find-in-directory path))))
         (root-length (length root-path))
         (rel-path (substring path root-length)))
    rel-path))

(defun acg/magit-open-remote-dwim (&optional dir)
  "Opens a remote repo URL in the exact DIR location. Prompts the user to choose a remote."
  (interactive)
  (let* ((file (buffer-file-name))
         (dir (expand-file-name (or dir default-directory)))
         (remote-name (magit-read-remote "Choose remote repository" nil t))
         (url (acg/parse-git-remote-url
               (magit-get "remote" remote-name "url")))
         ;; Identify remote hosting type (e.g., GitHub, GitLab, etc.)
         (remote-hosting-type
          (cond
           ((string-match-p (regexp-quote "github") url) "GitHub")
           ((string-match-p (regexp-quote "gitlab") url) "GitLab")
           (t (completing-read "Choose remote hosting type:"
                               '("GitHub" "GitLab"))))))
    ;; Adapt to GitLab's URL standard
    (when (string= remote-hosting-type "GitLab")
      (setq url (concat url "/-")))
    ;; Adapt to differences in file and directory URL paths
    (if file
        (setq url (concat url "/blob/" (magit-get-current-branch)
                          "/" (acg/project-get-root-relative-path file)))
      (setq url (concat url "/tree/" (magit-get-current-branch)
                        "/" (acg/project-get-root-relative-path dir))))
    ;; Perform actions
    (browse-url url)
    (message "Opening %s" url)))

(setq ob-mermaid-cli-path "/usr/local/bin/mmdc")

(setq org-directory my/org-base-dir)

(setq org-hide-emphasis-markers t)

(setq org-duration-format 'h:mm)

(setq org-log-done 'time)

(setq org-todo-keywords
      '((sequence "TODO" "DOING" "BLOCKED" "|" "DONE" "ARCHIVED")))

(setq org-refile-targets '((concat my/org-base-dir "inbox.org" :maxlevel . 2)))

(after! org
  (setq org-tags-match-list-sublevels 'indented))

(after! org
  (setq org-tags-column -120))

(after! org
  (setq org-agenda-prefix-format
        '((agenda . " %i %-12:c%?-12t% s")
          (todo   . " ")
          (tags   . " %i %-12:c")
          (search . " %i %-12:c"))))

(after! org
  (setq org-image-actual-width (/ (display-pixel-width) 2)))

(defun clockable-files ()
  "Returns org files that contain clockable items"
  (directory-files-recursively (concat my/org-base-dir "journal/") ".org"))

(defun convert-org-clocktable-time-to-hhmm (time-string)
  "Converts a time string to HH:MM"
  (if (> (length time-string) 0)
      (progn
        (let* ((s (s-replace "*" "" time-string))
               (splits (split-string s ":"))
               (hours (car splits))
               (minutes (car (last splits)))
               )
          (if (= (length hours) 1)
              (format "0%s:%s" hours minutes)
            (format "%s:%s" hours minutes))))
    time-string))

(setq org-agenda-hide-tags-regexp ".")

(after! org
  (setq org-agenda-files (list
                          (concat my/org-base-dir "inbox.org")
                          (concat my/org-base-dir "agenda.org"))))

(after! org
  (setq calendar-week-start-day 1)
  (setq calendar-intermonth-text
        '(propertize
          (format "%2d"
                  (car
                   (calendar-iso-from-absolute
                    (calendar-absolute-from-gregorian (list month day year)))))
          'font-lock-face 'font-lock-warning-face))
  (setq calendar-intermonth-header
        (propertize "Wk"                  ; or e.g. "KW" in Germany
                    'font-lock-face 'font-lock-keyword-face)))

(defun org-hugo-new-subtree-post-capture-template ()
  "Returns `org-capture' template string for new Hugo post.
   See `org-capture-templates' for more information."
  (let* ((title (read-from-minibuffer "Post Title: ")) ;Prompt to enter the post title
         (fname (org-hugo-slug title)))
    (mapconcat #'identity
               `(
                 ,(concat "* TODO " title)
                 ":PROPERTIES:"
                 ,(concat ":EXPORT_FILE_NAME: " fname)
                 ":END:"
                 "%?\n")          ;Place the cursor here finally
               "\n")))

(after! org
  ;; clean the intial existing template list
  (setq org-capture-templates '())


  ;; add inbox capture template
  (add-to-list 'org-capture-templates
               `("i" "Inbox" entry (file "inbox.org")
                 ,(concat "* TODO %?\n"
                          "/Entered on/ %U")))

  ;; add meeting capture template
  (add-to-list 'org-capture-templates
               `("m" "Meeting" entry (file+headline "agenda.org" "Planned")
                 ,(concat "* %? :meeting:\n"
                          "<%<%Y-%m-%d %a %H:00>>")))

  ;; add hugo capture template
  (add-to-list 'org-capture-templates
               '("h"                ;`org-capture' binding + h
                 "Hugo post"
                 entry
                 ;; It is assumed that below file is present in `org-directory'
                 ;; and that it has a "Blog Ideas" heading. It can even be a
                 ;; symlink pointing to the actual location of all-posts.org!
                 (file+olp "~/Dropbox/notes/org/blog.org" "Posts")
                 (function org-hugo-new-subtree-post-capture-template) :prepend t)))

(setq org-babel-clojure-backend 'cider)

(setq org-superstar-headline-bullets-list '("⁖" "◉" "○" "✸" "✿"))

(setq org-journal-dir (concat my/org-base-dir "journal"))

(setq org-journal-date-prefix "#+TITLE: ")

(setq org-journal-time-prefix "* ")

(setq org-journal-date-format "%a, %Y-%m-%d")

(setq org-journal-file-format "%Y-%m-%d.org")

(setq org-journal-enable-agenda-integration t)

(setq org-journal-tag-persistent-alist '(
                                         ("@work" . ?w)
                                         ("@private" . ?p)
                                         ("fontys" . ?f)
                                         ("ctoprogram" . ?c)
                                         ("integrationlab" . ?i)
                                         ("swat" . ?s)
                                         ("interruption" . ?!)
                                         ("various" . ?v)
                                         ))

(setq org-roam-directory (concat my/org-base-dir "roam"))

(setq +org-roam-open-buffer-on-find-file nil)

;; https://github.com/emacsorphanage/zoom-window
(custom-set-variables
 '(zoom-window-mode-line-color "purple"))
(map! :desc "Zoom/Unzoom the current window"
      :n "C-w z" 'zoom-window-zoom)

(after! smartparens
  (add-hook! (clojure-mode emacs-lisp-mode lisp-mode cider-repl-mode) :append #'smartparens-strict-mode))
(map!
 :map smartparens-mode-map
 ;; smartparens maps (navigation ops)
 :nvie "C-M-f" #'sp-forward-sexp
 :nvie "C-M-b" #'sp-backward-sexp
 :nvie "C-M-u" #'sp-backward-up-sexp
 :nvie "C-M-d" #'sp-down-sexp
 ;; smartparens maps (split join slurp barf)
 :nie "M-s" #'sp-split-sexp
 :nie "M-j" #'sp-join-sexp
 :nvie "C->" #'sp-forward-slurp-sexp
 :nvie "C-<" #'sp-forward-barf-sexp
 :nvie "C-{" #'sp-backward-slurp-sexp
 :nvie "C-}" #'sp-backward-barf-sexp)

(after! css-mode
  (setq css-indent-offset 2))

(setq json-reformat:indent-width 2)

(map! :desc "JSON beautify"
      :n "C-c C-f" 'json-mode-beautify)

(map! :localleader
      :map markdown-mode-map
      :prefix ("i" . "Insert")
      :desc "Blockquote"    "q" 'markdown-insert-blockquote
      :desc "Bold"          "b" 'markdown-insert-bold
      :desc "Code"          "c" 'markdown-insert-code
      :desc "Emphasis"      "e" 'markdown-insert-italic
      :desc "Footnote"      "f" 'markdown-insert-footnote
      :desc "Code Block"    "s" 'markdown-insert-gfm-code-block
      :desc "Image"         "i" 'markdown-insert-image
      :desc "Link"          "l" 'markdown-insert-link
      :desc "List Item"     "n" 'markdown-insert-list-item
      :desc "Pre"           "p" 'markdown-insert-pre
      (:prefix ("h" . "Headings")
        :desc "One"   "1" 'markdown-insert-header-atx-1
        :desc "Two"   "2" 'markdown-insert-header-atx-2
        :desc "Three" "3" 'markdown-insert-header-atx-3
        :desc "Four"  "4" 'markdown-insert-header-atx-4
        :desc "Five"  "5" 'markdown-insert-header-atx-5
        :desc "Six"   "6" 'markdown-insert-header-atx-6))

(after! cider
  (set-popup-rule! "^\\*cider-repl" :side 'right :size 0.3 :quit nil :ttl 0))
